core/src/com/unciv/logic/city/CityConstructions.kt

```diff
@@ -293,9 +295,17 @@ class CityConstructions {
         builtBuildings.remove(buildingName)
     }
 
-    fun purchaseConstruction(constructionName: String) {
+    fun purchaseConstruction(constructionName: String, cityScreen: CityScreen?) {
+        if (!getConstruction(constructionName).postBuildEvent(this)) {
+            if (cityScreen != null) // it can be null, because AI does not care about popups
```

Comment:
There's a mixup here between UI and models
Purchase Construction is a pure model function, and should not know UI
If you want to conditionally popup, this function should return a boolean

Reply:
That's a good point. Thank you for valuable review.